### PR TITLE
Fix a bug in indenting multiline comments with tabs (issue #348)

### PR DESCRIPTION
**Pull request by takumif**
Fixed the issue described in #348 in which the lines after the first in a block comment are stripped of leading spaces and are aligned incorrectly

### CODE REVIEW

File: src/vs/editor/common/commands/shiftCommand.ts [hunk 1]
```diff
@@ -71,6 +71,8 @@ export class ShiftCommand implements EditorCommon.ICommand {
 			}
 		}
 
+		var isMultilineComment = false;
```
Comment (bashrc-real) [inline src/vs/editor/common/commands/shiftCommand.ts]: [minor] Declare near first usage

File: src/vs/editor/common/commands/shiftCommand.ts [hunk 2]
```diff
@@ -99,7 +105,23 @@ export class ShiftCommand implements EditorCommon.ICommand {
...
+				// Add space to align the multiline comment
+				desiredIndent += " ";
+			}
+
+			// Check whether we're in a multiline comment or not
+			if (lineText.length >= 2 + indentationEndIndex) {
+				if (lineText.indexOf("/*") === indentationEndIndex) {
+					isMultilineComment = true;
+				}
+				if (lineText.indexOf("*/") !== -1) {
```
Comment (bashrc-real) [inline src/vs/editor/common/commands/shiftCommand.ts]: [minor] You can pass the start index in the call to indexOf. Will reduce atleast 2 comparisons.

Comment (bgourlie) [inline src/vs/editor/common/commands/shiftCommand.ts]: Hardcoding language-specific comment delimiters here seems wrong.

**Comment by takumif** [comment]
@bgourlie I agree that applying this regardless of the language in use isn't desirable.

I'm thinking whether it'd be better to change the indentation behavior entirely - right now, (un)indenting causes lines to be aligned at multiples of tabs and extra whitespaces are removed. On the other hand editors like Sublime Text retain the relative positioning by simply inserting tabs in front of lines, regardless of the content. I feel that'd make more sense, and it wouldn't cause problems like #348.

**Comment by alexandrudima** [comment]
The initial implementation of indenting/unindenting did exactly what @takumif describes. It would insert a leading `\t` or leading n spaces and it would remove a leading `\t` or leading n spaces. The code was also much simpler :smile: 

The feedback we got then was also not great, as sometimes you'd end up with mixed indent whitespace on the same line. Some examples:
 * pasting some code that is indented with spaces when your file is indented with tabs. The only way to "fix" the indentation was to either unindent all the way and indent blocks manually or to play with the find/replace. Now I just indent once and unindent once. Maybe in the future we can add an option to always do this on paste -- some sort of "fixIndentOnPaste".
 * a line that is "misaligned" by one space by accident (indentation 5 chars) would drive me crazy, I had to go to it and explicitly remove the extra space. Now I just indent once and unindent once over a large block of code.

When I think about it, my personal intent when I indent/undindent is for the code to "go to the right place". When using spaces at 4 chars for example, the "right place" for code is never at 5. However, as #348 shows, the "right place" for JSDoc style comments is sometimes at 5 chars.

I haven't read in detail the code, I would prefer an approach where indenting/unindenting tries its best to maintain whitespace diffs between lines in block comments.

**Comment by alexandrudima** [comment]
Fixed #348 with 53186e97a62d059d8bc6bcffd88896ea7a531571 that leaves the ShiftCommand language agnostic.

**Pull request closed without merging**